### PR TITLE
RDM-2013 Do not override case data's default classification on each e…

### PR DESCRIPTION
**Pull request by andrewfolga**
…vent

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-2013

### Change description ###
Actual: Every time event is triggered all of case's default data classification gets its fields reset from case type definition even though it has already been set during previously triggered events.

Expected: We should only set default security classification once so that already existing fields do not get their classification reset on each new event and only new fields do.

Notes:
- performance improvement in that we do not need to loop over all case fields once we found the searched for one,
- fixes some tests that were treating Document type classification as complex object whereas it should be simple type

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseDataService.java [hunk 1]
```diff
@@ -49,13 +51,19 @@ private void deduceDefaultClassifications(final JsonNode dataNode,
                     final String caseFieldType = caseField.getFieldType().getType();
                     if (caseFieldType.equalsIgnoreCase(COMPLEX_TYPE)) {
                         found = true;
-                        deduceClassificationForComplexType(dataNode, fieldIdPrefix, fieldName, caseField);
+                        deduceClassificationForComplexType(dataNode,
+                                                           getExistingDataClassificationNodeOrEmpty(existingDataClassificationNode, fieldName), fieldIdPrefix, fieldName, caseField);
+                        break;
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseDataService.java]: Performance improvement in that we do not need to loop over all case fields once we found the searched for one.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java [hunk 2]
```diff
@@ -85,4 +86,9 @@ public FieldType getCollectionFieldType() {
     public void setCollectionFieldType(FieldType collectionFieldType) {
         this.collectionFieldType = collectionFieldType;
     }
+
+    @Override
+    public String toString() {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java]: do we invoke this toString often? if yes, reflection can hurt performance

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseDataService.java [hunk 3]
```diff
@@ -31,13 +31,15 @@
     private static final String CLASSIFICATION = "classification";
 
     public Map<String, JsonNode> getDefaultSecurityClassifications(final CaseType caseType,
-                                                                   final Map<String, JsonNode> caseData) {
-        final JsonNode dataClassification = cloneAndConvertDataMap(caseData);
-        deduceDefaultClassifications(dataClassification, caseType.getCaseFields(), EMPTY_STRING);
-        return MAPPER.convertValue(dataClassification, STRING_JSON_MAP);
+                                                                   final Map<String, JsonNode> caseData,
+                                                                   final Map<String, JsonNode> currentDataClassification) {
+        final JsonNode newDataClassification = cloneAndConvertDataMap(caseData);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseDataService.java]: perhaps better name it clonedDataClassification?

File: src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java [hunk 4]
```diff
@@ -1328,7 +1322,6 @@ public void shouldReturn201WhenPostCreateCaseEventWithNoDataForCitizen() throws
         assertEquals(savedCaseDetails.getData(), caseAuditEvent.getData());
         assertEquals(SUMMARY, caseAuditEvent.getSummary());
         assertEquals(DESCRIPTION, caseAuditEvent.getDescription());
-        assertEquals(mapper.convertValue(EXPECTED_CLASSIFICATION, new TypeReference<HashMap<String, JsonNode>>() {}), caseAuditEvent.getDataClassification());
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java]: Readd it.

File: src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java [hunk 5]
```diff
@@ -1245,7 +1239,6 @@ public void shouldReturn201WhenPostCreateCaseEventWithNoDataForCaseworker() thro
...
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java]: Readd it.

File: src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java [hunk 6]
```diff
@@ -3261,7 +3251,6 @@ private void shouldReturn201WhenPostCreateCaseEventWithValidData(String userRole
...
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/std/CaseDetailsEndpointIT.java]: Readd it.

**Pull request merged**